server/src/main/java/org/candlepin/async/JobManager.java

```diff
@@ -1086,8 +1087,18 @@ private AsyncJobStatus processJobFailure(AsyncJobStatus status, EventSink eventS
         // Rollback any unsent events
         eventSink.rollback();
 
-        // Only store the exception type and the message. Luckily for us, toString does exactly that.
-        String result = throwable != null ? throwable.toString() : null;
+        Object result = null;
+        if (throwable != null) {
+            // CandlepinException subclasses can override the message() method to enhance the message from a
+            // simple String to something with more useful information, like ImportConflictException does
+            // to include specific import conflicts that happened.
+            if (throwable.getCause() instanceof CandlepinException) {
+                result = ((CandlepinException) throwable.getCause()).message();
```

Comment:
@Ceiu @kahowell Regarding your previous [comment](https://github.com/candlepin/candlepin/pull/2331#issuecomment-495802031) about reducing the exception result to a message string, and the potential for using a "failure" class, it turns out that we already have a mechanism for subclasses of `CandlepinException` to enhance their "message" to be more than a single string. This is currently used only by `ImportConflictException`, in order to add a list of "conflicts" that can happen during an import failure, which are surely useful to the client.

From my point of view, this is enough for now, until we need anything more complex in the future.

Reply:
Maybe. For some context, CandlepinExceptions are generally used for exceptions to be passed out of the resources directly to the end-user via RESTeasy. 

I'll take a look at this a bit closer during review.

Reply:
Looking at the replaced code, this may work, but it's likely going to be extra work to get the message. The replaced line which calls .toString will, by default, give us a string equivalent to the following:
```String msg = throwable.getClass().getName() + ": " + throwable.getLocalizedMessage()```

The .getLocalizedMessage call will call into .getMessage(), which in the case of CandlepinException, will return the same thing as .message().getDisplayMessage()/.message().toString(). Basically, we get similar (identical?) output without needing the additional checks here.

Reply:
Also, if we ever want to standardize this handling a bit more, this gist [1] is something I had in the prototyping stage that was eventually cut due to it being a similar amount of boilerplate that throwable's default .toString already does for us.

[1] https://gist.github.com/Ceiu/b401efa25b1eda7637958b17eef8bf4b

Reply:
Unfortunately, the replaced code does not work, because the "message" returned by the `.toString()` method is a simple String which contains only the message of the exception (it resides on `Throwable.java` as such: `private String detailMessage;`), while CandlepinException has a completely different, custom `message` field which is a: `protected final ExceptionMessage message;`. In the case of the `ImportConflictException`, this is further extended as `ConflictExceptionMessage`, which, when serialized looks like this:
```
{
  "result" :
   {
    "displayMessage": "Import is the same as existing data",
    "requestUuid": "ff8080816b08d246016b08d2f35f0087",
    "conflicts": [ "MANIFEST_SAME", "DISTRIBUTOR_CONFLICT" ]
   }
}
```
, while the replaced code returns:
```
{
  "result" : "org.candlepin.sync.ImportConflictException: Import is the same as existing data"
}
```
CandlepinException does not can cannot override `getMessage()`, because the signature returns a simple String. This is why `message()` is used instead.

Reply:
Argh... okay. And because we're setting the exception, the result is getting translated away from that. I wonder if the special handling of the CandlepinException is something we should be handling specifically in this job, since it seems to be unique to it.

Reply:
Took a quick look through the other pinsetter jobs. Only the following the HealEntireOrgJob and HypervisorUpdateJob use CandlepinExceptions, and even those are both BadRequestExceptions which can easily be replaced by a generic JobExecutionException. So, this logic seems to be something unique to the ImportJob. At least, that's all I saw with a surface-level glance through the jobs.

The difficult problem to address here is how to both set the result and throw an exception to indicate the job status in a way that lets us control both without some convoluted wrapper or output scheme. We used a special listener in Pinsetter to run code similar to what you've added to JobManager, but I feel like we can do better. Let me think on this a bit.

Reply:
Since this is a unique case, this is what I've come up with:
* Leave the `JobManager` set the result as it originally was (`String result = throwable != null ? throwable.toString() : null;`). This works fine currently with all exceptions that jobs are throwing, except `ImportConflictException`.
* Have ImportJob transform the `ImportConflictException` to a new, [`ImportConflictJobException` (which extends `JobExecutionException`)](https://gist.github.com/nikosmoum/49dad9e7673245a42deedecb402b45ba) which provides the list of conflicts as a result of its `.toString()`. This way
  - We are being compatible with what `JobManager` does for all other exceptions.
  - We propagate a `JobExecutionException`, not a `CandlepinException`
  - We save the valuable part of the data (the list of conflicts and display message) in the result, without having the useless/unfit for purpose data(requestUuid & REST code).

Reply:
I'm okay with this for now.

We still have the interesting problem of resolving what to do when we want to set a result while throwing an exception. When I get back, I'll look into a kind of hybrid approach between what you have here and the ExceptionMessage idea from CandlepinException. That might work well enough to cover all of these type of cases in the future.